### ISSUE TITLE
Plugin: camera - crash when app resumed or screen unlocked if controller is disposed

### ISSUE THREAD

**Issue by mattiagalati** [flutter/flutter#18348]
## Steps to Reproduce

1. Start example in camera repository at https://github.com/flutter/plugins/tree/master/packages/camera/example
2. Edit example to pretend there is a previous page that navigate to camera main page:
```dart
class PreHome extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new Container(
      child: new RaisedButton(
        onPressed: () {
          Navigator.of(context).push(
            new MaterialPageRoute(
              builder: (BuildContext context) {
                return new CameraExampleHome();
              },
            ),
          );
        },
        child: const Text("CAMERA"),
      ),
    );
  }
}
```
3. Dispose the camera controller when the CameraExampleHome is disposed (this was taken by 0.2.0 examples of this plugin, I cannot find it in the new example, will explain this later)
```dart
  @override
  void dispose() {
    try{
      controller?.dispose();
    }catch(err){
      print(err);
    }
    super.dispose();
  }
```
4. Run example, tap on "CAMERA" button, and then proceede to select a camera, starting the preview (no need to take a picture actually)
5. Using software or hardware buttons, go back to the first page with only the button "CAMERA" and lock screen or pause app
6. Unlock screen or resume app. The app crashes

## Logs
```
[        ] Synced 0.7MB.
[   +2 ms] _flutter.listViews: {}
[  +20 ms] Connected to _flutterView/0x7f741e8818.
[        ] 🔥  To hot reload your app on the fly, press "r". To restart the app entirely, press "R".
[        ] An Observatory debugger and profiler on Lenovo P2a42 is available at: http://127.0.0.1:8109/
[        ] For a more detailed help message, press "h". To quit, press "q".
[+3528 ms] I/FlutterActivityDelegate( 6503): onResume setting current activity to this
[+5389 ms] V/BoostFramework( 6503): mAcquireFunc method = public int com.qualcomm.qti.Performance.perfLockAcquire(int,int[])
[        ] V/BoostFramework( 6503): mReleaseFunc method = public int com.qualcomm.qti.Performance.perfLockRelease()
[        ] V/BoostFramework( 6503): mAcquireTouchFunc method = public int com.qualcomm.qti.Performance.perfLockAcquireTouch(android.view.MotionEvent,android.util.DisplayMetrics,int,int[])
[        ] V/BoostFramework( 6503): mIOPStart method = public int com.qualcomm.qti.Performance.perfIOPrefetchStart(int,java.lang.String)
[        ] V/BoostFramework( 6503): mIOPStop method = public int com.qualcomm.qti.Performance.perfIOPrefetchStop()
[   +2 ms] V/BoostFramework( 6503): BoostFramework() : mPerf = com.qualcomm.qti.Performance@35464e0
[+3388 ms] I/CameraManager( 6503): Using legacy camera HAL.
[ +138 ms] I/CameraDeviceState( 6503): Legacy camera service transitioning to state CONFIGURING
[   +1 ms] I/RequestThread-0( 6503): Configure outputs: 2 surfaces configured.
[        ] D/Camera  ( 6503): app passed NULL surface
[   +4 ms] I/RequestThread-0( 6503): configureOutputs - set take picture size to 4160x3120
[  +19 ms] I/CameraDeviceState( 6503): Legacy camera service transitioning to state IDLE
[   +4 ms] I/RequestQueue( 6503): Repeating capture request set.
[   +4 ms] W/LegacyRequestMapper( 6503): convertRequestMetadata - control.awbRegions setting is not supported, ignoring value
[        ] W/LegacyRequestMapper( 6503): Only received metering rectangles with weight 0.
[        ] W/LegacyRequestMapper( 6503): Only received metering rectangles with weight 0.
[ +372 ms] I/CameraDeviceState( 6503): Legacy camera service transitioning to state CAPTURING
[+2642 ms] I/art     ( 6503): Do partial code cache collection, code=21KB, data=25KB
[        ] I/art     ( 6503): After code cache collection, code=21KB, data=25KB
[        ] I/art     ( 6503): Increasing code cache capacity to 128KB
[+1251 ms] I/RequestQueue( 6503): Repeating capture request cancelled.
[  +34 ms] E/BufferQueueProducer( 6503): [SurfaceTexture-1-6503-1] queueBuffer: BufferQueue has been abandoned
[  +10 ms] E/BufferQueueProducer( 6503): [SurfaceTexture-1-6503-1] cancelBuffer: BufferQueue has been abandoned
[        ] E/BufferQueueProducer( 6503): [SurfaceTexture-1-6503-1] cancelBuffer: BufferQueue has been abandoned
[        ] E/BufferQueueProducer( 6503): [SurfaceTexture-1-6503-1] cancelBuffer: BufferQueue has been abandoned
[+6877 ms] I/CameraManager( 6503): Using legacy camera HAL.
[  +80 ms] I/FlutterActivityDelegate( 6503): onResume setting current activity to this
[  +20 ms] E/GLConsumer( 6503): [SurfaceTexture-0-6503-0] setDefaultBufferSize: GLConsumer is abandoned!
[        ] E/BufferQueueProducer( 6503): [SurfaceTexture-0-6503-0] query: BufferQueue has been abandoned
[   +1 ms] E/Legacy-CameraDevice-JNI( 6503): LegacyCameraDevice_nativeDetectSurfaceDimens: Error while querying surface width No such device (-19).
[        ] D/AndroidRuntime( 6503): Shutting down VM
[        ] E/AndroidRuntime( 6503): FATAL EXCEPTION: main
[        ] E/AndroidRuntime( 6503): Process: io.flutter.plugins.cameraexample, PID: 6503
[        ] E/AndroidRuntime( 6503): java.lang.IllegalArgumentException: Surface was abandoned
[        ] E/AndroidRuntime( 6503): 	at android.hardware.camera2.utils.SurfaceUtils.getSurfaceSize(SurfaceUtils.java:70)
[        ] E/AndroidRuntime( 6503): 	at android.hardware.camera2.params.OutputConfiguration.<init>(OutputConfiguration.java:184)
[        ] E/AndroidRuntime( 6503): 	at android.hardware.camera2.params.OutputConfiguration.<init>(OutputConfiguration.java:94)
[        ] E/AndroidRuntime( 6503): 	at android.hardware.camera2.impl.CameraDeviceImpl.createCaptureSession(CameraDeviceImpl.java:489)
[        ] E/AndroidRuntime( 6503): 	at io.flutter.plugins.camera.CameraPlugin$Camera.startPreview(CameraPlugin.java:672)
[        ] E/AndroidRuntime( 6503): 	at io.flutter.plugins.camera.CameraPlugin$Camera.access$2000(CameraPlugin.java:237)
[        ] E/AndroidRuntime( 6503): 	at io.flutter.plugins.camera.CameraPlugin$Camera$3.onOpened(CameraPlugin.java:437)
[        ] E/AndroidRuntime( 6503): 	at android.hardware.camera2.impl.CameraDeviceImpl$1.run(CameraDeviceImpl.java:134)
[        ] E/AndroidRuntime( 6503): 	at android.os.Handler.handleCallback(Handler.java:751)
[        ] E/AndroidRuntime( 6503): 	at android.os.Handler.dispatchMessage(Handler.java:95)
[        ] E/AndroidRuntime( 6503): 	at android.os.Looper.loop(Looper.java:154)
[        ] E/AndroidRuntime( 6503): 	at android.app.ActivityThread.main(ActivityThread.java:6111)
[        ] E/AndroidRuntime( 6503): 	at java.lang.reflect.Method.invoke(Native Method)
[        ] E/AndroidRuntime( 6503): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
[        ] E/AndroidRuntime( 6503): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
[        ] E/AndroidRuntime( 6503): Caused by: android.hardware.camera2.legacy.LegacyExceptionUtils$BufferQueueAbandonedException
[        ] E/AndroidRuntime( 6503): 	at android.hardware.camera2.legacy.LegacyExceptionUtils.throwOnError(LegacyExceptionUtils.java:73)
[        ] E/AndroidRuntime( 6503): 	at android.hardware.camera2.legacy.LegacyCameraDevice.getSurfaceSize(LegacyCameraDevice.java:585)
[        ] E/AndroidRuntime( 6503): 	at android.hardware.camera2.utils.SurfaceUtils.getSurfaceSize(SurfaceUtils.java:68)
[        ] E/AndroidRuntime( 6503): 	... 14 more
```
```
```

## Flutter Analyze
```
Analyzing example...                                             
No issues found! (ran in 1.9s)
```

## Flutter Doctor
```
[✓] Flutter (Channel beta, v0.4.4, on Linux, locale it_IT.UTF-8)
    • Flutter version 0.4.4 at /home/adaron/flutter
    • Framework revision f9bb4289e9 (4 settimane fa), 2018-05-11 21:44:54 -0700
    • Engine revision 06afdfe54e
    • Dart version 2.0.0-dev.54.0.flutter-46ab040e58

[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at /home/adaron/.android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-27, build-tools 27.0.3
    • Java binary at: /usr/local/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b01)
    • All Android licenses accepted.

[✓] Android Studio (version 3.0)
    • Android Studio at /usr/local/android-studio
    • Flutter plugin version 23.2.1
    • Dart plugin version 171.4424
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b01)

[✓] VS Code (version 1.23.1)
    • VS Code at /usr/share/code
    • Dart Code extension version 2.13.2

[✓] Connected devices (1 available)
    • Lenovo P2a42 • ZY223WWWVD • android-arm64 • Android 7.0 (API 24)

• No issues found!
```

## Notes
- The problem descripted above seems not occur on android 5.0: I encountered it only on devices with Android 7.0 by now.
- Consider that my workflow is: RECORD LIST > START CAMERA and TAKE PICTURE > FILL A FORM and SAVE. Theese 3 routes are tied by an await Navigator which inform the list if user has abandoned or completed the flow, in order to refresh the list. This information is passed through POP method of navigator.
- When I started developing my app, the plugin was in version 0.2.0 which has a "_controller?.dispose()" call inside the "dispose" method of the parent StatelessWidget. There is no similar call in new example, so I wonder if in my case is still needed, but using the 0.2.0 version, I noticed that removing the "_controller?.dispose()" in my case (maybe for the workflow above) lead to overheating of the camera.
- I tried to push a further route after the picture is taken (in the attached file is a route where I show the picture in full width), when the screen locked or app paused, nothing happens at resume until I go back to the route with the button "CAMERA".
[main.dart.zip](https://github.com/flutter/flutter/files/2088270/main.dart.zip)

**Comment by zoechi** [comment]
@Hixie

**Comment by kelostrada** [comment]
Just wanted to drop my 2 cents - I reverted to `0.4.3` and it works as intended so probably something in engine changed meanwhile which broke camera. Also I want to point out that what I found out that later (after 0.4.4) the camera stops working at all: #18329

**Comment by mattiagalati** [comment]
Just for your information, on a test device equipped with Android Oreo (Samsung Galaxy S7), avoiding to dispose the camera controller, causes the system notification to user that the app is using too much resources in background. Don't remember the exact quote, but I can check if needed and provide a screenshot.

**Comment by zoechi** [comment]
The exact error message would definitely be helpful, but it might be better to create a new issue for that.

**Comment by wiltwitt** [comment]
I do have an iphone 6 which is 4 years old but camera widget/page crashes when clicked and really slow to load image selected from library.  I will try to post the error if I can but this is really disappointing.   I gave Flutter about 4 months try so I got used to things but this is too slow.  Oh man, do I have to go back to Xamarin???

**Comment by kmorkos** [comment]
I've identified the bug and submitted a pull request here: https://github.com/flutter/plugins/pull/645

**Comment by wheel1992** [comment]
This issue is quite urgent. Will it get merged soon?

**Comment by kmorkos** [comment]
@wheel1992 I believe we're waiting on my CLA to be approved.

In the meantime, you can use the patch by referencing my branch directly. You would add the dependency to your pubspec.yaml like so:

```yaml
camera:
  git:
    url: git://github.com/kmorkos/plugins.git
    ref: camera/crash-fix
    path: packages/camera/
```

**Comment by maral** [comment]
I had to combine both @kmorkos and @CurosMJ solutions (but I wrapped the whole startPreview method in a try-catch block, since also the other calls were raising exceptions) and never had any more exception. Here is the fork, if anyone is interested: https://github.com/maral/plugins/tree/camera/crash-fix

**Comment by d-silveira** [comment]
This is a major show stopper, it should not remain open since July, without hot-fix update, come on guys, I used @maral 's workaround to save me from bining a major product.

**Comment by rlee1990** [comment]
Any update on when this fix will be released?

**Comment by kmorkos** [comment]
@rlee1990 it was released as v0.2.4 a few days ago.

**Comment by rlee1990** [comment]
Thanks @kmorkos I didn't see it on https://pub.dartlang.org/ I will update my app.

**Comment by rlee1990** [comment]
@kmorkos I get this error when I try to use it
`Because flutter_crushn depends on camera ^0.2.4 which doesn't match any versions, version solving failed.`

**Comment by zoechi** [comment]
@rlee1990 there is a dependency in your project that is not compatible with other dependencies in your project or with the Flutter/Dart version you are using

**Comment by kmorkos** [comment]
It appears the update hasn't been pushed to pub yet. If you track the master branch you'll have the fix.

**Comment by mattiagalati** [comment]
At the moment I'm using the fork version which does not causes crash.
I will release a new app within 3 days in which I will test the new version of pugin. Will reply soon.
Many thanks.

**Comment by zoechi** [comment]
Without additional information, we are unfortunately not sure how to resolve this issue. We are therefore reluctantly going to close this bug for now. Please don't hesitate to comment on the bug if you have any more information for us; we will reopen it right away!
Thanks for your contribution.

**Comment by jclink** [comment]
Not sure if this has been resolved; I recently encountered the surface exception using camera: ^0.4.3+2.  I was able to avoid it by limiting my calls to controller.dispose().  In short, I restructured it so that controller.dispose() is called only once, by the enclosing state widget dispose() method.  In other areas, when check if controller == null, I returned instead of disposing.  My suspicion being a timing offset from asynchronous operations in my code.

In my project, this avoids the surface issue, but it should be noted that I am only using the back camera. Also, I am not recording video.

**Comment by shabeenabarde** [comment]
> Not sure if this has been resolved; I recently encountered the surface exception using camera: ^0.4.3+2. I was able to avoid it by limiting my calls to controller.dispose(). In short, I restructured it so that controller.dispose() is called only once, by the enclosing state widget dispose() method. In other areas, when check if controller == null, I returned instead of disposing. My suspicion being a timing offset from asynchronous operations in my code.
> 
> In my project, this avoids the surface issue, but it should be noted that I am only using the back camera. Also, I am not recording video.

Can you provide code upon how you have initiated this?
